lift/Readme.md

```diff
@@ -6,271 +6,360 @@
+          │                          │           └─────────┼▶│
+          │                          │      migration      │ │
+          │        migration         │          ┌──────────┤ │
+          │            ┌─────────────│◀─────────┴──────────┼─┤
+          │◀───────────┘             │                     │ │
+          │                          │                     │ ▼
+          ▼                          ▼                     ▼
+```
 
-## Detailed design
+### Lift Client
 
-As mentioned in the [basic example](#basic-example), a simple `prisma migrate` is needed to migrate the database and create a new migration.
+The Lift Client is built into the Prisma 2 CLI but is also accessible programmatically in the Prisma SDK. The Lift Client has 3 methods:
 
-These are the commands:
+#### Save
 
-```bash
-prisma migrate                                Creates a migration, clears the draft and migrates the database
+Save method takes the difference of your schema file with the current state of your datasources and creates a migration folder. This migration folder contains
+the steps to sync your datasources with your schema file. The template of the migration folder is `${current_timestamp}-${migration-name}/`
 
-       migrate plan                           Creates a migration without applying it
-       migrate apply                          Applies all unapplied migrations
-       migrate draft                          Pushes the new changes into a draft without creating a migration
-       migrate rollback                       Rolls back migration(s)
+```
+migrations/
+  └─ 20190920142118-initial/
+    └─ steps.json
+    └─ schema.prisma
+    └─ README.md
 ```
 
-Let's discuss a few questions.
+A migration contains 3 files:
 
-### Naming of migration folders
+- **steps.json:** contains a JSON list of steps to run against the database
+- **schema.prisma:** contains a snapshot of your `schema.prisma` file at a specific point in time.
+- **README.md:** contains information about the migration. Includes the underlying raw commands (e.g. SQL) that run against the datasource.
 
-When executing `prisma migrate` without providing any args, a migration with the name `yyyyMMddHHmmss` (e.g. `20190322092247`) will be created. Especially when
-working in a team, it can be very useful to add a custom name to the migration. A name can be provided like this:
+**Note:** `migrate save` doesn't run migrations, it simply creates them.
 
-```bash
-$ prisma migrate --name my-initial-migration
+#### Up
+
+Up runs your outstanding migrations against the datasources, synchronizing the datasources with your schema file. Up depends on the migrations generated by the
+Save method.
+
+Up runs the unapplied `up` scripts in ascending order by the migration's timestamp. For example, if we have the following migrations directory:
+
+```
+migrations/
+  └─ 20190920142118-initial/
+    └─ steps.json
+    └─ schema.prisma
+    └─ README.md
+    └─ before.sh
```

Comment:
`before.sh` and `after.sh` are a new concept here (or did I miss it?). I assume those are hooks? If so, and these are the common names for them, this should be explained in the "Hooks" sections.

Reply:
good call, it's defined now